File: lib/services/score_calculator.dart
import '../models/app_settings.dart';

class ScoreCalculator {
  const ScoreCalculator();

  double scoreForCalories(double calories, AppSettings settings) {
    final points = [...settings.curve]..sort((a, b) => a.calories.compareTo(b.calories));
    if (calories <= points.first.calories) return points.first.score;
    if (calories >= points.last.calories) return points.last.score;
    for (var i = 0; i < points.length - 1; i++) {
      final a = points[i];
      final b = points[i + 1];
      if (calories >= a.calories && calories <= b.calories) {
        final t = (calories - a.calories) / (b.calories - a.calories);
        return (a.score + t * (b.score - a.score)).clamp(0, 100).toDouble();
      }
    }
    return 0;
  }

  String label(double score) => score >= 95 ? 'perfect' : score >= 85 ? 'good' : score >= 80 ? 'okay' : 'bad';
}

File: app/src/main/java/com/sam/caloriestreak/domain/calculation/ScoreCalculator.kt
package com.sam.caloriestreak.domain.calculation

import kotlin.math.abs

data class ScorePoint(val calories: Double, val score: Double)

class ScoreCalculator(points: List<ScorePoint> = defaultPoints) {
    private val curve = points.sortedBy { it.calories }
    private val targetPoint = curve.maxByOrNull { it.score }

    fun calculate(calories: Double): Double {
        require(curve.size >= 2) { "At least two score points are required" }
        targetPoint?.let { target ->
            if (abs(calories - target.calories) <= TARGET_EPSILON) return 100.0
        }
        if (calories <= curve.first().calories) return curve.first().score.coerceIn(0.0, 100.0)
        if (calories >= curve.last().calories) return curve.last().score.coerceIn(0.0, 100.0)
        val pair = curve.zipWithNext().first { (left, right) -> calories in left.calories..right.calories }
        val ratio = (calories - pair.first.calories) / (pair.second.calories - pair.first.calories)
        return (pair.first.score + ratio * (pair.second.score - pair.first.score)).coerceIn(0.0, 100.0)
    }

    fun status(score: Double): String = when {
        score < 40.0 -> "Bad"
        score < 80.0 -> "Poor"
        score < 85.0 -> "Streak safe"
        score < 95.0 -> "Good"
        else -> "Excellent"
    }

    companion object {
        const val DEFAULT_TARGET = 1650.0
        private const val TARGET_EPSILON = 0.000001

        val defaultPoints = pointsForTarget(DEFAULT_TARGET)

        fun pointsForTarget(target: Double): List<ScorePoint> {
            require(target in 800.0..5000.0) { "Target must be between 800 and 5000 kcal" }
            val scale = target / DEFAULT_TARGET
            return listOf(
                ScorePoint(800.0 * scale, 0.0),
                ScorePoint(1200.0 * scale, 40.0),
                ScorePoint(1400.0 * scale, 80.0),
                ScorePoint(target, 100.0),
                ScorePoint(1800.0 * scale, 75.0),
                ScorePoint(2000.0 * scale, 20.0),
                ScorePoint(2200.0 * scale, 0.0)
            )
        }

        fun forTarget(target: Double) = ScoreCalculator(pointsForTarget(target))
    }
}
